src/genetic.c

```diff
@@ -0,0 +1,451 @@
+ * UINT16_MAX
+ *
+ * @param population This array stores all the values of the population and will
+ * be filled
+ */
+static void fillTable(uint16_t population[]) {
+
+  unsigned int populationArraySize = populationSize * dimensions;
+
+  for (unsigned int i = 0; i < populationArraySize; i++) {
+
+    population[i] = linear_congruential_random_generator() * UINT16_MAX;
+  }
+}
+
+/**
+ * This function is used to choose two parents based off of a tourney approach
+ *
+ * @param populationStrength  an array that stores all the fitness of the
+ * population
+ * @param firstParentIndex a return parameter for storing the index of the first
+ * chosen parent
+ * @param secondParentIndex a return parameter for storing the index of the
+ * second chosen parent
+ */
+static void tourney(float populationStrength[], unsigned int* firstParentIndex,
+                    unsigned int* secondParentIndex) {
+
+  float chosenIndexes[populationSize];
+  float bestFitness, secondbestFitness = 0;
+
+  for (unsigned int i = 0; i < tournamentSelectionsSize; i++) {
+
+    unsigned int index =
+        (unsigned int)(linear_congruential_random_generator() * populationSize);
+    unsigned int isAlreadyChosen = 1;
+
+    chosenIndexes[i] = index;
+
+    if (i != 0) {
+
+      for (unsigned int j = 0; j < i; j++) {
+
+        if (chosenIndexes[j] == index) {
+          i--;
+          isAlreadyChosen = 0;
+          break;
+        }
+      }
+    }
+
+    float fitness = populationStrength[index];
+
+    if (!i || (isAlreadyChosen && bestFitness > fitness)) {
+
+      *secondParentIndex = *firstParentIndex;
+
+      *firstParentIndex = index;
+      bestFitness = fitness;
+
+    } else if (i == 1 || secondbestFitness > fitness) {
+      *secondParentIndex = index;
+      secondbestFitness = fitness;
+    }
+  }
+}
+
+/**
+ * This function simulates the mutation genetic operation on our created child
+ *
+ * @param child  this char array represents the created child and at each index
+ * is a digit
+ * @param childLength the length of the created child
+ */
+
+static void mutate(char child[], const unsigned int childLength) {
+
+  float mutation =
+      (unsigned int)(linear_congruential_random_generator() * populationSize) %
+      childLength;
+
+  if (mutation <= mutationRate) {
+
+    int mutatedIndex = (unsigned int)(linear_congruential_random_generator() *
+                                      populationSize) %
+                       childLength;
+    int newValue = (unsigned int)(linear_congruential_random_generator() *
+                                  populationSize) %
+                   10;
+    // We handle the last digit
+    if (mutatedIndex % UINT16_Digit_Count == 0) {
+      newValue = (unsigned int)(linear_congruential_random_generator() *
+                                populationSize) %
+                 7;
+    }
+    // Converts int to char
+    child[mutatedIndex] = newValue + '0';
```

Comment:
The child is a string and the newValue is an int between 0-9. I add that value to the ASCII value of '0' to get the appropriate char for the string.

Reply:
Oh I see, what you're trying to do here. I think it would be cleaner to define a macro or a function named toAscii(int) which does what you're doing here

Reply:
I agree that a macro would be good here

Reply:
In fact, an inline function would be even better.